Add transfer operator S3 to (generic) SQL

maggesssss opened the pull request:
closes: #23666

This PR includes a new Transfer Operator that reads a CSV File from S3 Storage and loads it into an existing Table of a generic SQL Database

I used csv.reader to read the file and `insert_rows` method of the existing DbApiHook.
Due to the fact that csv.reader is not reading the complete file into the memory, also large files can be loaded somehow efficiently.

I am happy for any feedback.

maggesssss committed:
added JSON linter to connection edit / add UI for field extra. On connection edit screen, existing extra data will be displayed idented

maggesssss committed:
Merge branch 'apache:main' into feat_conn_ui

Commit:
Released new S3 to SQL Transfer Operator Including Documentation and Tests

Commit:
added missing sql_conn_id parameter to example DAG to make it more clear and explicit

Commit:
fixed non-unique task_ids and operator variable names in S3 to Sql example DAG

Taragolis commented on airflow/providers/amazon/aws/transfers/s3_to_sql.py:
```diff
@@ -0,0 +1,158 @@
+except ImportError as e:
+    from airflow.exceptions import AirflowOptionalProviderFeatureException
+
+    raise AirflowOptionalProviderFeatureException from e
```
[`csv`](https://docs.python.org/3/library/csv.html) is a builtin Python library

Taragolis commented on airflow/providers/amazon/aws/transfers/s3_to_sql.py:
```diff
@@ -0,0 +1,158 @@
+
+        This operator downloads a file from an S3, reads it via `csv.reader`
+        and inserts the data into a SQL database using `insert_rows` method.
+        All SQL hooks are supported, as long as it is of type DbApiHook
```
Hmmm... Operator limited by CSV sources?

vincbeck replied:
Agree with this. Either you explicitly specify in the operator name that this operator reads only CSV, either you can make this operator generic by adding a new parameter `parser` which would be a function responsible of reading the input from the source. The actual processing of CSV in your would be handled by this new function `parser`. Between the 2 solutions, I prefer the second one. My 2 cents

maggesssss replied:
@vincbeck I agree. I like the Idea of using a `parser` parameter. Do you think it makes sense to use a default parser?

vincbeck replied:
I dont think it makes sense as default but I do think that providing at least one parser in code makes sense. A CSV parser would be excellent!

maggesssss replied:
Hi @vincbeck, I have changed the operator to use a parser instead and have documented it accordingly. 
Could you please have a and let me know what you think about it

vincbeck replied:
It looks really good! Thanks for making the changes. I added some comments but they are mostly nitpicks and minor comments. Good job!

Taragolis commented on airflow/providers/amazon/aws/transfers/s3_to_sql.py:
```diff
@@ -0,0 +1,158 @@
+
+        s3_hook = S3Hook(aws_conn_id=self.aws_conn_id)
+        self._file = s3_hook.download_file(key=self.s3_key, bucket_name=self.s3_bucket)
+
```
No need to define `self._file` use regular variable:
- This only use in one method
- This attribute not defined in class constructor

Taragolis commented on airflow/providers/amazon/aws/transfers/s3_to_sql.py:
```diff
@@ -0,0 +1,158 @@
+        try:
+            # open with newline='' as recommended
+            # https://docs.python.org/3/library/csv.html#csv.reader
+            with open(self._file, newline="") as file:
```
Better use [NamedTemporaryFile](https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile)

maggesssss replied:
Done

vincbeck commented on airflow/providers/amazon/aws/example_dags/example_s3_to_sql.py:
```diff
@@ -0,0 +1,84 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Following up on the discussion here #22438, this file should be moved in `tests/system/providers/amazon/aws/example_s3.py`

maggesssss replied:
@vincbeck I have modified the example DAG according to AIP-47, but I was not able to test it yet. Will do it as soon as I have fixed my breeze environment

vincbeck replied:
The system test looks really good! Thanks for the effort to have converted the example DAG!

maggesssss replied:
@vincbeck I added a SQL Check operator to count the rows inserted. Tests were running fine locally.
Can you give me a hint which conn_id's I have to use?
For S3 topics I guess `aws_default`, but for generic sql? `sql_default`?

vincbeck commented on airflow/providers/amazon/aws/transfers/s3_to_sql.py:
```diff
@@ -0,0 +1,158 @@
+        s3_hook = S3Hook(aws_conn_id=self.aws_conn_id)
+        self._file = s3_hook.download_file(key=self.s3_key, bucket_name=self.s3_bucket)
+
+        hook = self._get_hook()
```
This operator uses 2 different hook, would be better to have a more specific name

```suggestion
        db_hook = self._get_hook()
```

vincbeck commented on airflow/providers/amazon/aws/transfers/s3_to_sql.py:
```diff
@@ -0,0 +1,158 @@
+            # Remove file downloaded from s3 to be idempotent.
+            os.remove(self._file)
+
+    def _get_hook(self) -> DbApiHook:
```
nit. You can decorate this function with @cached_property

josh-fell replied:
Perhaps this doesn't need to be a property at all since it's only used in the `execute()` method?

vincbeck replied:
There is some work going on to standardize the hook access in Amazon provider package. See https://github.com/apache/airflow/pull/29001. I agree with you it is not necessary to store the hook in a property but (and this is only my personal opinion), using @cached_property makes the code cleaner

maggesssss replied:
@vincbeck I have pushed some changes, please let me know if it's fine now

vincbeck commented on docs/apache-airflow-providers-amazon/operators/transfer/s3_to_sql.rst:
```diff
@@ -0,0 +1,75 @@
+
+============================
+Amazon S3 to SQL
+============================
```
```suggestion
==============
Amazon S3 to SQL
==============
```

maggesssss replied:
This results in an error when building the docs:

```
------------------------------ Error   6 --------------------
 WARNING: Title overline too short.

File path: apache-airflow-providers-amazon/operators/transfer/s3_to_sql.rst (18)

  13 |     "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  14 |     KIND, either express or implied.  See the License for the
  15 |     specific language governing permissions and limitations
  16 |     under the License.
  17 | 
> 18 | ==============
  19 | Amazon S3 to SQL
  20 | ==============
  21 | 
  22 | Use the ``S3ToSqlOperator`` transfer to copy data from an Amazon Simple Storage Service (S3) file into an existing
  23 | SQL table.
------------------------------ Error   7 --------------------
 WARNING: Title underline too short.

File path: apache-airflow-providers-amazon/operators/transfer/s3_to_sql.rst (37)

  32 | ---------
  33 | 
  34 | .. _howto/operator:S3ToSqlOperator:
  35 | 
  36 | Amazon S3 To SQL Transfer Operator
> 37 | ============================
  38 | 
  39 | To get more information about this operator visit:
  40 | :class:`~airflow.providers.amazon.aws.transfers.s3_to_sql.S3ToSqlOperator`
  41 | 
  42 | Example usage:

############################## End docs build errors summary ##############################



```

vincbeck replied:
Oops, the diff must be non very reliable for this then. Just be sure, the length of `==============` is the same as the text

vincbeck commented on docs/apache-airflow-providers-amazon/operators/transfer/s3_to_sql.rst:
```diff
@@ -0,0 +1,75 @@
+.. _howto/operator:S3ToSqlOperator:
+
+Amazon S3 To SQL Transfer Operator
+==============================================
```
```suggestion
Amazon S3 To SQL Transfer Operator
============================
```

vincbeck commented on docs/apache-airflow-providers-amazon/operators/transfer/s3_to_sql.rst:
```diff
@@ -0,0 +1,75 @@
+
+
+
+
```
```suggestion



```

josh-fell commented on airflow/providers/amazon/aws/transfers/s3_to_sql.py:
```diff
@@ -0,0 +1,158 @@
+if TYPE_CHECKING:
+    from airflow.utils.context import Context
+
+from typing_extensions import Literal
```
```suggestion
from airflow.typing_compat import Literal
```
Let's us the compat module instead here.

josh-fell commented on airflow/providers/amazon/aws/transfers/s3_to_sql.py:
```diff
@@ -0,0 +1,158 @@
+        if csv_reader_kwargs:
+            self.csv_reader_kwargs = csv_reader_kwargs
+        else:
+            self.csv_reader_kwargs = {}
```
```suggestion
        self.csv_reader_kwargs = csv_reader_kwargs or {}
```
Perhaps a nit, but seems more readable.

josh-fell commented on docs/apache-airflow-providers-amazon/operators/transfer/s3_to_sql.rst:
```diff
@@ -0,0 +1,75 @@
+
+Use the ``S3ToSqlOperator`` transfer to copy data from an Amazon Simple Storage Service (S3) file into an existing
+SQL table.
+Only CSV Format is supported.
```
Should the operator fail fast and prior to downloading from S3 if the file is not a CSV? Can this be implied by the `s3_key` value?

If you agree, then let's add a test for this situation too.

josh-fell commented on docs/apache-airflow-providers-amazon/operators/transfer/s3_to_sql.rst:
```diff
@@ -0,0 +1,75 @@
+
+
+
+You can also pass **column_list='inferred** if you want the operator to read the column names from the first row of the CSV File:
```
```suggestion
You can also pass ``column_list='infer'`` if you want the operator to read the column names from the first row of the CSV File:
```

maggesssss committed:
Merge branch 'apache:main' into feat_s3_to_sql_transfer

Commit:
save WIP

Commit:
save WIP

Commit:
Merge branch 'backup' into feat_s3_to_sql_transfer

hussein-awala committed:
Skip DockerOperator task when it returns a provided exit code (#28996)

potiuk committed:
Clarify that we are using lowest supported version for static checks (#29024)

There is a good reason static checks always use lowest supported
Python version. This PR clarifies that it happens and why.

maggesssss committed:
Added parser
parameter which allows the user to add a custom parser.
Example parser added to docs

removed following args:
- csv_reader_kwargs
- skip_first_row
- column_list "infer" option
These arguments are not working with a customer parser at the moment

Changed to NamedTempoaryFile

Added s3_hook.get_key before downloading to check if file exists

Updated test and docs

maggesssss committed:
WIP: migrated example DAG to system tests

maggesssss committed:
removed example DAG (migrated to system tests)

vincbeck commented on airflow/providers/amazon/aws/transfers/s3_to_sql.py:
```diff
@@ -0,0 +1,127 @@
+    """
+        Loads Data from S3 into a SQL Database.
+        You need to provide a parser function that takes a filename as an input
+        and returns a iterable of rows
```
```suggestion
        and returns an iterable of rows
```

vincbeck commented on airflow/providers/amazon/aws/transfers/s3_to_sql.py:
```diff
@@ -0,0 +1,127 @@
+        e.g. to use a CSV parser that yields rows line-by-line, pass the following
+        function:
+
+        def parse_csv(filepath):
```
Love it!

vincbeck commented on tests/providers/amazon/aws/transfers/test_s3_to_sql.py:
```diff
@@ -0,0 +1,102 @@
+    @patch("airflow.providers.amazon.aws.transfers.s3_to_sql.NamedTemporaryFile")
+    @patch("airflow.providers.amazon.aws.transfers.s3_to_sql.DbApiHook.insert_rows")
+    @patch("airflow.providers.amazon.aws.transfers.s3_to_sql.S3Hook.get_key")
+    def test_execute(self, mock_get_key, mock_insert_rows, mock_tempfile, mock_parser):
```
For the sake to have a better coverage, could you add a test when the hook does not have the method `insert_rows` defined?

vincbeck commented on tests/system/providers/amazon/aws/example_s3_to_sql.py:
```diff
@@ -0,0 +1,173 @@
+    )
+
+    @task(trigger_rule=TriggerRule.ONE_FAILED, retries=0)
+    def watcher():
```
This task is already defined in `tests/system/utils/watcher.py`. No need to redefine it, you can just import it

maggesssss committed:
Update airflow/providers/amazon/aws/transfers/s3_to_sql.py

Co-authored-by: Vincent <97131062+vincbeck@users.noreply.github.com>

pierrejeambrun committed:
Update owner list for release dockerhub image workflow (#29023)

maggesssss committed:
changed internal method _get_hook
to cached property db_hook

maggesssss committed:
removed unused imports
use imported watcher task

SoxMax committed:
Update how PythonSensor returns values from python_callable (#28932)

* Update how PythonSensor returns values from python_callable

* test poke returns the xcom value

* update test to only run poke

* reformat based on changes

* use full if rather than ternary

yxiao1996 committed:
Update gantt chart UI to display queued state of tasks (#28686)

* Update gantt chart UI to display queued state of tasks

* Fix style issue raise by static check

* Add comments to lines related to drawing boxes in gantt chart

Co-authored-by: eladkal <45845474+eladkal@users.noreply.github.com>

o-nikolas committed:
Patch MockExecutor in tests instead of "prod" code (#29028)

Patch the MockExecutor into the executor list within test code rather
than core/prod code.

IAL32 committed:
Add log for AWS Glue Job Console URL (#28925)

* Log AWS Glue Job Console URL

* Added GlueJobLogsLink

o-nikolas commented on airflow/providers/amazon/aws/transfers/s3_to_sql.py:
```diff
@@ -0,0 +1,127 @@
+        hook = conn.get_hook()
+        if not callable(getattr(hook, "insert_rows", None)):
+            raise AirflowException(
+                "This hook is not supported. The hook class must have  an `insert_rows` method."
```
```suggestion
                "This hook is not supported. The hook class must have an `insert_rows` method."
```

o-nikolas commented on tests/providers/amazon/aws/transfers/test_s3_to_sql.py:
```diff
@@ -0,0 +1,102 @@
+        db.merge_conn(
+            models.Connection(
+                conn_id="s3_test",
+                conn_type="s3",
```
Isn't the `s3` `conn_type` deprecated (in #25980)?

jedcunningham committed:
Add norm to triage group (#29029)

josh-fell committed:
Fix `project-name` generation for Providers `test-type` in Breeze (#29030)

The `breeze testing tests --test-type "Providers[...]"` command would fail for an invalid `project-name` for `docker compose` if the provider(s) specified required a dotted path (e.g. `microsoft.azure`, `dbt.cloud`, etc.).

This fixes the `project-name` generation in Breeze when using the `Providers` test type for those dotted-path providers.

josh-fell committed:
Fix params rendering in AzureSynapseHook Python API docs (#29041)

Trivial update to add a newline between the description and `:param:` directives in the AzureSynapseHook docstring. This allows the parameters to be rendered properly in the Python API documentation for better readability.

potiuk committed:
Pin ruff to specific version and prevent from "fixing" cli conftest (#29040)

New Ruff release started to "fix" our cli conftest cofiguration,
leading to mypy complaining about the changed line.

This change prevents ruff from "fixing" the line as well as pins
ruff to latest released version so that any future changes are
applied deliberately when we upgrade.

potiuk committed:
Sanitize url_for arguments before they are passed (#29039)

The url_for of flask has special arguments that start with `_` and we
should sanitize the ones that come with the request before passing them.

vchiapaikeo committed:
Add support for write_on_empty in BaseSQLToGCSOperator (#28959)

maggesssss committed:
changed conn type from s3 to aws

maggesssss committed:
Update airflow/providers/amazon/aws/transfers/s3_to_sql.py

Co-authored-by: Niko Oliveira <onikolas@amazon.com>

maggesssss committed:
added sql_default conn_id
for SqlExecuteQueryOperators

maggesssss committed:
added seek(0) to tempfile after it was
downloaded

maggesssss committed:
converted sample_data to raw
string and added

SQLTableCheckOperator to check if
lines have been successfully imported

maggesssss committed:
added test for bad hook
(without insert_rows method)
and optimized db_hook property

vincbeck reviewed (approved):
LGTM! Thanks for making the changes

Taragolis commented on airflow/providers/amazon/aws/transfers/s3_to_sql.py:
```diff
@@ -0,0 +1,128 @@
+        with NamedTemporaryFile() as local_tempfile:
+
+            s3_obj.download_fileobj(local_tempfile)
+            local_tempfile.seek(0)
```
```suggestion
            s3_obj.download_fileobj(local_tempfile)
            local_tempfile.flush()
            local_tempfile.seek(0)
```

I've better add explicit call flush, not every functions/methods which work with file buffers call it.

mhenc committed:
AIP-44 Initialize methods map for Internal API RPC endpoint in the method (#28841)

vandonr-amz committed:
uniformize getting hook through cached property in aws sensors (#29001)

uniformize getting hook through cached property in aws sensors

RachitSharma2001 committed:
Update provide_bucket_name() decorator to handle new conn_type (#28706)

* Update provide_bucket_name() decorator to handle new conn_type

* Utilize caching

* Add documentation

vandonr-amz committed:
nice table at the end of system test execs (#29051)

BasPH committed:
Move project and license docs down in menu to start with developer-focused docs (#28956)

BasPH committed:
Annotate and simplify code samples in DAGs doc (#29027)

potiuk committed:
Fix kerberos authentication for the REST API. (#29054)

Previously we assigned kerberos user name directly to the flask user,
but this had no chance to work because we expect FAB user there and
our security code crash with 'str' has no attribute 'perms'.

This PR uses Kerberos username (including the Kerberos realm) to
retrieve the user from the security manager. This means that
the user name has to have the form of `user_name@KERBEROS_REALM`.

The reason why we are not using email (despite similarities of
the realm and domain name is that those are often different. Email
domain names have often nothing to do the with the realms within
organisations, and it seems safer to put fully qualified names
including the realm in order to uniquely identify the users in
case the organisation uses more than one REALM.

Fixes: #28919

Co-authored-by: BMFH <bogner85@mail.ru>

Co-authored-by: BMFH <bogner85@mail.ru>

Taragolis committed:
Migrate Helm tests to `pytest` (#29063)

surabathini committed:
logging poke info when external dag is not none and task_id and task_ids are none (#28097)

mhenc committed:
AIP-44: Add CLI command for running standalone Internal API (#28425)

potiuk committed:
Fix pre-commit warning for exclude in inclusive-language check (#29057)

The exclude regexp contained /* which was not really what we meant
(no harm done but latest pre-commit started to warn about it).

vincbeck committed:
AIP-44 Migrate DagModel.get_paused_dag_ids to Internal API (#28693)

Taragolis committed:
Renaming nose compatible methods in flavour of regular pytest naming (#29035)

BasPH committed:
Capitalize dag to DAG (#29064)

potiuk committed:
Fix SFTP operator's template fields processing (#29068)

The SFTP operator had logic in `__init__` that was running checks
on passed arguments, but that means that the operator woudl fail
when using TaskFlow.

Fixes: #27328

bharat99k committed:
fixing import error for dataset (#29007)

ephraimbuddy committed:
Properly test that audit log secrets are masked for connection objects (#29011)

Previous test was wrong as the connection was updated in place which returns None

uranusjr committed:
Minor improvements to serde helpers (#28978)

MrGeorgeOwl committed:
Add deferrable mode for Big Query Transfer operator (#27833)

VladaZakharova committed:
Add deferrable mode to DataFusionStartPipelineOperator (#28690)

snjypl committed:
Fix minor typo in breeze example command. (#29072)

hussein-awala committed:
Emit DataDog statsd metrics with metadata tags (#28961)

Bowrna committed:
listener plugin example added (#27905)

Taragolis committed:
Refactor `TestLocalTaskJob.test_process_sigterm_works_with_retries` (#28990)

Taragolis committed:
Mark `test_process_sigterm_works_with_retries` quarantined again (#29075)

snjypl committed:
FTP operator has logic in __init__ (#29073)

Move the logic from __init__ to executor for FTP operator

snjypl committed:
Upgrade bitnami/postgresql helm depenency from 10.5.3 to 12.1.9 (#29071)

This version of the chart uses different variable names for setting usernames and passwords in the postgres database.
`postgresql.auth.enablePostgresUser` is used to determine if the "postgres" admin account will be created.
`postgresql.auth.postgresPassword` sets the password for the "postgres" user.
`postgresql.auth.username` and `postrgesql.auth.password` are used to set credentials for a non-admin account if desired.
`postgresql.postgresqlUsername` and `postgresql.postresqlPassword`, which were used in the previous version of the chart, are no longer used.
Users will need to change these variable names in their values files if they are using the helm chart.

Co-authored-by: Caleb Woofenden <caleb.woofenden@bitsighttech.com>

Aakcht committed:
Chart: add doc note about podtemplate images (#29032)

DrFaust92 committed:
bump (#28602)

bbovenzi committed:
Check for run_id url param when linking to graph/gantt views (#29066)

hussein-awala committed:
emit dagrun failed duration when timeout (#29076)

snjypl committed:
Migrate DagFileProcessorManager.clear_nonexistent_import_errors to Internal API (#28976)

maggesssss committed:
removed type-hint from property db_hook.
As long as the hook has an insert_rows method, this operator can work.

maggesssss committed:
fixed import sorting

maggesssss committed:
added flush() to
local_tempfile after download is finished

maggesssss committed:
removed unneccessary
self._get_hook()
db_hook is now a cached property

maggesssss committed:
Merge branch 'apache:main' into feat_s3_to_sql_transfer
